internal/core/docker_context.go

```diff
@@ -0,0 +1,328 @@
+	contextType   typeGetter
+	endpointTypes map[string]typeGetter
+}
+
+// newConfig creates a config object
+func newConfig(contextType typeGetter, endpoints ...namedTypeGetter) contextConfig {
+	res := contextConfig{
+		contextType:   contextType,
+		endpointTypes: make(map[string]typeGetter),
+	}
+
+	for _, e := range endpoints {
+		res.endpointTypes[e.name] = e.typeGetter
+	}
+	return res
+}
+
+// metadata contains metadata about a context and its endpoints
+type metadata struct {
+	Name      string         `json:",omitempty"`
+	Metadata  any            `json:",omitempty"`
+	Endpoints map[string]any `json:",omitempty"`
+}
+
+func (s *metadataStore) contextDir(id contextdir) string {
+	return filepath.Join(s.root, string(id))
+}
+
+type untypedContextMetadata struct {
+	Metadata  json.RawMessage            `json:"metadata,omitempty"`
+	Endpoints map[string]json.RawMessage `json:"endpoints,omitempty"`
+	Name      string                     `json:"name,omitempty"`
+}
+
+func (s *metadataStore) getByID(id contextdir) (metadata, error) {
+	fileName := filepath.Join(s.contextDir(id), metaFile)
+	bytes, err := os.ReadFile(fileName)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return metadata{}, errdefs.NotFound(fmt.Errorf("context not found: %w", err))
+		}
+		return metadata{}, err
+	}
+
+	var untyped untypedContextMetadata
+	r := metadata{
+		Endpoints: make(map[string]any),
+	}
+
+	if err := json.Unmarshal(bytes, &untyped); err != nil {
+		return metadata{}, fmt.Errorf("parsing %s: %w", fileName, err)
+	}
+
+	r.Name = untyped.Name
+	if r.Metadata, err = parseTypedOrMap(untyped.Metadata, s.config.contextType); err != nil {
+		return metadata{}, fmt.Errorf("parsing %s: %w", fileName, err)
+	}
+
+	for k, v := range untyped.Endpoints {
+		if r.Endpoints[k], err = parseTypedOrMap(v, s.config.endpointTypes[k]); err != nil {
+			return metadata{}, fmt.Errorf("parsing %s: %w", fileName, err)
+		}
+	}
+
+	return r, err
+}
+
+// list returns a list of all Docker contexts
+func (s *metadataStore) list() ([]metadata, error) {
+	ctxDirs, err := listRecursivelyMetadataDirs(s.root)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	res := make([]metadata, 0, len(ctxDirs))
+	for _, dir := range ctxDirs {
+		c, err := s.getByID(contextdir(dir))
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				continue
+			}
+			return nil, fmt.Errorf("read metadata: %w", err)
+		}
+		res = append(res, c)
+	}
+
+	return res, nil
+}
+
+type contextdir string
+
+func isContextDir(path string) bool {
+	s, err := os.Stat(filepath.Join(path, metaFile))
+	if err != nil {
+		return false
+	}
+
+	return !s.IsDir()
+}
+
+func listRecursivelyMetadataDirs(root string) ([]string, error) {
+	fis, err := os.ReadDir(root)
+	if err != nil {
+		return nil, err
+	}
+
+	var result []string
+	for _, fi := range fis {
+		if fi.IsDir() {
+			if isContextDir(filepath.Join(root, fi.Name())) {
```

Comment:
question: Is this a duplicate check as if we start with root and fi is a directory it should not be possible that this could ever fail?